internal/host_test.go

```diff
@@ -0,0 +1,113 @@
+package internal_test
+
+import (
+	"context"
+	"github.com/tetratelabs/wazero"
+	"github.com/tetratelabs/wazero/api"
+	"testing"
+)
+
+// instantiateWapcHost instantiates a test waPC host and returns it and a cleanup function.
+func instantiateWapcHost(t *testing.T, r wazero.Runtime) (*wapcHost, api.Closer) {
+	h := &wapcHost{t: t}
+	// Export host functions (in the order defined in https://wapc.io/docs/spec/#required-host-exports)
+	if host, err := r.NewModuleBuilder("wapc").
+		ExportFunction("__host_call", h.hostCall,
+			"__host_call", "bind_ptr", "bind_len", "ns_ptr", "ns_len", "cmd_ptr", "cmd_len", "payload_ptr", "payload_len").
+		ExportFunction("__console_log", h.consoleLog,
+			"__console_log", "ptr", "len").
+		ExportFunction("__guest_request", h.guestRequest,
+			"__guest_request", "op_ptr", "ptr").
+		ExportFunction("__host_response", h.hostResponse,
+			"__host_response", "ptr").
+		ExportFunction("__host_response_len", h.hostResponseLen,
+			"__host_response_len").
+		ExportFunction("__guest_response", h.guestResponse,
+			"__guest_response", "ptr", "len").
+		ExportFunction("__guest_error", h.guestError,
+			"__guest_error", "ptr", "len").
+		ExportFunction("__host_error", h.hostError,
+			"__host_error", "ptr").
+		ExportFunction("__host_error_len", h.hostErrorLen,
+			"__host_error_len").
+		Instantiate(testCtx, r); err != nil {
+		t.Errorf("Error instantiating waPC host - %v", err)
+		return h, nil
+	} else {
+		return h, host
+	}
+}
+
+type wapcHost struct {
+	t                  *testing.T
+	consoleLogMessages []string
+}
+
+// hostCall is the WebAssembly function export "__host_call", which initiates a host using the callHandler using
+// parameters read from linear memory (wasm.Memory).
+func (w *wapcHost) hostCall(ctx context.Context, m api.Module, bindPtr, bindLen, nsPtr, nsLen, cmdPtr, cmdLen, payloadPtr, payloadLen uint32) int32 {
+	panic("TODO")
```

Comment:
intentionally didn't backfill impl for all the things as it would make review burden quite hard on first go, as well the impl.